examples/her/train_dqn_bit_flip.py

```diff
@@ -0,0 +1,248 @@
+import argparse
+
+import gym
+import gym.spaces as spaces
+import numpy as np
+import torch
+import torch.nn as nn
+
+from pfrl.q_functions import DiscreteActionValueHead
+from pfrl import agents
+from pfrl import experiments
+from pfrl import explorers
+from pfrl import utils
+from pfrl import replay_buffers
+
+from pfrl.initializers import init_chainer_default
+
+
+def reward_fn(dg, ag):
+    return -1.0 if (ag != dg).any() else 0.0
+
+
+class BitFlip(gym.GoalEnv):
+    """BitFlip environment from https://arxiv.org/pdf/1707.01495.pdf
+
+    Args:
+        n: State space is {0,1}^n
+    """
+
+    def __init__(self, n):
+        self.n = n
+        self.action_space = spaces.Discrete(n)
+        self.observation_space = spaces.Dict(
+            dict(
+                desired_goal=spaces.MultiBinary(n),
+                achieved_goal=spaces.MultiBinary(n),
+                observation=spaces.MultiBinary(n),
+            )
+        )
+        self.clear_statistics()
+
+    def compute_reward(self, achieved_goal, desired_goal, info):
+        return reward_fn(desired_goal, achieved_goal)
+
+    def _check_done(self):
+        success = (
+            self.observation["desired_goal"] == self.observation["achieved_goal"]
+        ).all()
+        return (self.steps >= self.n) or success, success
+
+    def step(self, action):
+        # Compute action outcome
+        bit_new = int(not self.observation["observation"][action])
+        new_obs = self.observation["observation"].copy()
+        new_obs[action] = bit_new
+        # Set new observation
+        dg = self.observation["desired_goal"]
+        self.observation["desired_goal"] = dg.copy()
+        self.observation["achieved_goal"] = new_obs
+        self.observation["observation"] = new_obs
```

Comment:
I had to create a new object here for training to work. The successes should start showing up in the logs within 1000 episodes.

```suggestion
        self.observation = {
            "desired_goal": dg.copy(),
            "achieved_goal": new_obs,
            "observation": new_obs,
        }
```

If I had to guess why this change is needed, it'd be because the `self.observation` object is being inserted into the replay buffer, and when we later change the object keys it updates for all the observations in the replay buffer because they share a source object? Maybe I'm wrong about the reason, but the new object is needed to train in my tests.

Reply:
Thanks! You're right, I think we either have to do this or return `self.observation.copy(), reward, done, {}` instead of `self.observation, reward, done, {}`